go/store/nbs/table_set.go

```diff
@@ -425,9 +425,63 @@ func (ts tableSet) flatten(ctx context.Context) (tableSet, error) {
 	return flattened, nil
 }
 
+// openForAdd will attempt to open every file named in |files| with the
+// table persister, returning a new chunkSourceSet for all of the files
+// if they are all able to be opened. An error will be returned if any
+// errors are encountered when opening the files.
+//
+// For any files which appear in |novel| or |upstream|, this function
+// will return clones of the chunk sources, instead of opening them
+// anew.
+func (ts tableSet) openForAdd(ctx context.Context, files map[hash.Hash]uint32, stats *Stats) (chunkSourceSet, error) {
+	ret := make(chunkSourceSet)
+	cleanup := func() {
+		for _, source := range ret {
+			source.close()
+		}
+	}
+	for h := range files {
+		if s, ok := ts.novel[h]; ok {
+			cloned, err := s.clone()
+			if err != nil {
+				cleanup()
+				return nil, err
+			}
+			ret[h] = cloned
+		} else if s, ok := ts.upstream[h]; ok {
+			cloned, err := s.clone()
+			if err != nil {
+				cleanup()
+				return nil, err
+			}
+			ret[h] = cloned
+		}
+	}
+	eg, ctx := errgroup.WithContext(ctx)
+	var mu sync.Mutex
+	for h, c := range files {
```

Comment:
I'm unfamiliar with the semantics here so it's not obvious what value c represents or how it feeds into the Open call below